### DISCUSSION TITLE
JS Path: Add a Clean Code "part 2" to supplement the one on Foundations

### DISCUSSION THREAD

**Discussion by impronen** [TheOdinProject/curriculum#25718] [Ideas]
I'm currently doing a rewrite to the Clean Code lesson in Foundations, as per [this feature request](https://github.com/TheOdinProject/curriculum/issues/23870).

The [Clean Code book](https://github.com/ryanmcdermott/clean-code-javascript)  has a lot of potentially very useful ideas that would greatly benefit the learners but the position of the Clean Code lesson prohibits talking about them. The lesson is placed before introduction of such basic building blocks of code like loops, array methods and object. This means that there is no use in giving advice about how passing objects as arguments can be useful, when the reader doesn't yet know what a object is. 

I have currently focused on trying to refactor a more robust article around basic concepts and with very simple examples, avoiding these more advanced topics. But they are very useful and I personally feel that I've gained a great deal reading about them. So I would love others to share this.

That is why I would think it a great thing to have a Clean Code part 2, perhaps somewhere around todo. This would give the lesson more breathing room as the learner is already familiar with lot more concepts and has had to do some pretty complex things. Using objects / destructuring as function parameters can be understood at this point, etc. 

I think this would be a relatively light lesson, something that wouldn't prolong the learning journey unduly but would still be beneficial to learners. If this is at some point converted to a Issue, I'd be interested in working on it, as I've had quite fun with the og lesson rewrite already

I'm curious to hear your thoughts.

**Comment by leokirasic** [comment]
I completely agree with you. Let's go for it.

**Reply by impronen** [reply]
So would the process for doing this be to open up an issue next with the requirements and then proceed with writing the lesson by just creating a new lesson file on a branch of the curriculum repo?

**Comment by CouchofTomato** [comment]
Yeah you can create a branch from a fork of the repo and add a new lesson, then you can create a PR so we can review it. Let me know if you need a bit more support around that.

Once agreed then to include it in the lesson content we need to make a seperate PR to the web app repo to include it but I can take that on if you aren't sure about how to do that.

**Reply by impronen** [reply]
Thanks @CouchofTomato for the tips! I'm sure I'll manage the fork/branch/pr side of things for the lesson. 

For the website repo I'd gladly leave it for someone else, as I've understood that it's working with RoR and I don't know any Ruby.

**Comment by nik-rev** [comment]
It seems that there has been no traction for a while here and I believe a clean code 2 lesson would be super beneficial.

I am happy to take this on, if the original poster doesn't mind? Just wanted to make sure everything is okay.

**Reply by impronen** [reply]
Hi! I've actually been working on a bit on this, just nothing complete yet - I have a half finished lesson of sorts. But lately I haven't had time to work on it at all, so if you want to work on this, I welcome the assistance!

The stuff I've done is on my local branch, I'll take a look at it later today after work and push it to github and share to you so you see what's already there. It's pretty rough on a lot of places and lot of stuff is still needed.

**Reply by nik-rev** [reply]
> Hi! I've actually been working on a bit on this, just nothing complete yet - I have a half finished lesson of sorts. But lately I haven't had time to work on it at all, so if you want to work on this, I welcome the assistance!
> 
> The stuff I've done is on my local branch, I'll take a look at it later today after work and push it to github and share to you so you see what's already there. It's pretty rough on a lot of places and lot of stuff is still needed.

Hey! That sounds great! I have an idea: what do you think about publishing your PR as a draft, so that me and others can easily work on it by "review" feature

**Reply by impronen** [reply]
Ah well, right now I'm rusty enough not to remember how to do that off the top of my head. I really should code more...

I'll review the contribution guide latest on Monday and make a PR draft after that. In the meantime, this is what I've come up with so far: https://github.com/impronen/odin-curriculum/blob/clean-code-2/javascript/javascript_in_the_real_world/clean_code_continued.md

**Reply by nik-rev** [reply]
> Ah well, right now I'm rusty enough not to remember how to do that off the top of my head. I really should code more...
> 
> I'll review the contribution guide latest on Monday and make a PR draft after that. In the meantime, this is what I've come up with so far: https://github.com/impronen/odin-curriculum/blob/clean-code-2/javascript/javascript_in_the_real_world/clean_code_continued.md

Nice, looks impressive. I'll give it a proper read once you submit it as a draft since then I can easily click on a "+" icon to add a suggestion

If you want to know how to make a draft PR, then it's simple really. 

1. Go to your fork of the curriculum
2. Switch to your branch where the lesson is
3. In the **< > Code** section, find **Contribute** button
4. Click "open a pull request"
5. Type in your information, create the pull request
6. On the right hand side, you should see the following:

```
Reviewers
No reviews
Still in progress? Convert to draft
```

7. Click "convert to draft"
